src/common/regions.ts

```diff
@@ -116,27 +121,65 @@ const counties: County[] = chain(state_county_map_dataset)
 class RegionDB {
   constructor(private regions: Region[]) {}
 
-  private filterByType(regionType: RegionType) {
-    return this.all().filter(region => region.regionType === regionType);
-  }
-
   findByFipsCode(fipsCode: FipsCode): Region | null {
     const region = this.regions.find(region => region.fipsCode === fipsCode);
     return region || null;
   }
 
+  findStateByUrlParams(stateId: string): State | null {
+    const foundState = this.states().find(
+      state =>
+        equalLower(state.urlSegment, stateId) ||
+        equalLower(state.stateCode, stateId),
+    );
+    return foundState || null;
+  }
+
+  findCountyByUrlParams(stateId: string, countyId: string): County | null {
+    const foundState = this.findStateByUrlParams(stateId);
+    if (!foundState) {
+      return null;
+    }
+
+    const stateCounties = this.counties().filter(
+      county => county.state.fipsCode === foundState.fipsCode,
+    );
+
+    const foundCounty = stateCounties.find(county =>
+      equalLower(county.urlSegment, countyId),
+    );
+
+    return foundCounty || null;
+  }
+
   all(): Region[] {
     return this.regions;
   }
 
-  states(): Region[] {
-    return this.filterByType(RegionType.STATE);
+  states(): State[] {
+    return this.all().filter(
+      (region): region is State => region.regionType === RegionType.STATE,
+    );
   }
 
-  counties(): Region[] {
-    return this.filterByType(RegionType.COUNTY);
+  counties(): County[] {
+    return this.all().filter(
+      (region): region is County => region.regionType === RegionType.COUNTY,
+    );
```

Comment:
Don't we already have `states` and `counties` constants defined in this file? Do we need to recalculate these?

FWIW- I really like to make method/function names reflect the performance characteristics.  I'd intuitively expect `RegionsDB.states()` to be a very cheap method call, but as-is it's going to iterate over 3000+ entries and construct a new list of the state entries, I think.  So we should either make sure the list is cached (so it _is_ cheap) or else name this `filterToStates()` or something to make it clearer that you shouldn't call it repeatedly.

Reply:
Updated, passing states and counties in the constructor, they can be accessed directly now (faster)